20190719 rrsets pagination

peterthomassen opened the pull request:
Functionality is there. What's missing:

- tests
- docs

peterthomassen commented:
Tests and documentation added.

To finalize this, I would like to discuss the appropriate response status code for requests that would require pagination, but don't use a `cursor` parameter. Currently, we produce `400` (and instruct how to use pagination). I wonder if there is a more specific code, such as

- `300 Multiple Choices`, but that's intended for varied versions of the same thing (e.g. languages)
- `403 Forbidden`, but very misleading
- `411 Length Required`, but refers to `Content-Length`
- `412 Precondition Failed` -- something with header conditions, but I don't fully understand
- `413 Request Entity Too Large`, but actually refers to the request payload, not the response
- `416 Requested Range Not Satisfiable`, but actually refers to `Range` headers

... any ideas?

peterthomassen commented:
Agreed with @nils-wisiol that status 400 is good enough.

No more open issues, ready for review.

nils-wisiol commented on api/desecapi/tests/test_rrsets.py:
```diff
@@ -42,15 +44,90 @@ def test_retrieve_my_rr_sets(self):
-        response = self.client.get_rr_sets(self.my_rr_set_domain.name)
+        response = self.client.get_rr_sets(self.my_rr_set_domain.name, query='?cursor=')
         self.assertStatus(response, status.HTTP_200_OK)
-        self.assertEqual(len(response.data), len(self._test_rr_sets()))
```
can we assert that length is `min(len(self...), page_size_setting)`?

peterthomassen replied:
sure! good solution, done

nils-wisiol commented on api/desecapi/tests/test_rrsets.py:
```diff
@@ -42,15 +44,90 @@ def test_retrieve_my_rr_sets(self):
+            [RRset(domain=self.my_domain, subname=str(i), ttl=123, type='A') for i in range(n)]
+        )
+
+        # No pagination
```
The following five blocks of code should be refactored into some sub-routine to make maintenance easier.

peterthomassen replied:
done! looks much better now, thx for the suggestion

nils-wisiol commented on api/desecapi/tests/test_rrsets.py:
```diff
@@ -344,6 +344,16 @@ def test_update_essential_properties(self):
+        response = self.client.patch(url, data)
+        self.assertStatus(response, status.HTTP_200_OK)
+        response = self.client.patch(url, data)
+        self.assertStatus(response, status.HTTP_200_OK)
```
Why patching twice?

peterthomassen replied:
I guess one should have been PUT, shall change

nils-wisiol commented on api/desecapi/serializers.py:
```diff
@@ -172,7 +172,7 @@ class RRsetSerializer(ConditionalExistenceModelSerializer):
     class Meta:
         model = RRset
-        fields = ('domain', 'subname', 'name', 'records', 'ttl', 'type',)
+        fields = ('created', 'domain', 'subname', 'name', 'records', 'ttl', 'type',)
```
This should also be added to the e2e schema test in `test/e2e/schemas.js`

peterthomassen replied:
good catch! done

nils-wisiol commented on docs/domains.rst:
```diff
@@ -141,10 +141,15 @@ The ``/api/v1/domains/`` endpoint reponds to ``GET`` requests with an array of
 
-to retrieve an overview of the domains you own.
+to retrieve an overview of the domains you own.  Domains are returned in
+reverse chronological order of their creation.
```
Does our code actually guarantee this? I am not aware of any sorting.

peterthomassen replied:
This is a new side-effect feature: The paginator paginates based on the created timestamp.

nils-wisiol commented on docs/rrsets.rst:
```diff
@@ -198,11 +198,32 @@ command::
 
-to retrieve the contents of a zone that you own.
+to retrieve the contents of a zone that you own.  RRsets are returned in
+reverse chronological order of their creation.
```
Same as above

peterthomassen replied:
This is a new side-effect feature: The paginator paginates based on the created timestamp.

peterthomassen commented:
> I am not sure what's the intention of [6e42035](https://github.com/desec-io/desec-stack/commit/6e42035ea01cc287e8b10aa40c0c44832399735e):
> 
> * it moves the default value of the `minimum_ttl` field, but as far as I can tell, this does not have any effect -- it's now just defined elsewhere. Could you enlighten me? blush

The intention was to solve the ghost migration issue that arouse when the value of `settings.MINIMUM_TTL_DEFAULT` changed. We solved it in the meantime with 0f1ed0ed523d87ad167cae831bb5cc7dc0f49485 (merged already), so I removed the fix from this PR.

> * it changes the test to `Domain.get` instead of `Domain.get_or_create`, which I confirmed is unnecessary to pass the tests.

That's true, but I figured the tests be as specific as possible. As I noticed that all tests create their domains before using them, I thought `get` would be the more clear route.

> Why does [6a849ea](https://github.com/desec-io/desec-stack/commit/6a849eaa1a96212d7835d3e52b0e47a36a2af677) increase the TTL?

Because the tests failed in my test environment which had the minimum TTL set to 3600 instead of 60. While I don't think that 3600 in particular needs to be supported, I figured that there's no point in overly low TTL values, just to decrease the likelihood of running into this kind of problem again in the future. (Think of someone else working on this, using a minimum TTL of 900, for example). That's not a perfect approach, of course, but I thought it would help a bit with test resilience.

I took all code comments into account, fixed things up, rebased, and I think we're good to go. Can you pls take another look? :blush:
